### ISSUE TITLE
Install dart_language_server if not already present

### ISSUE THREAD

**Issue by bradyt** [emacs-lsp/lsp-mode#541]
> Great! Another possibility is to use `(executable-find "dart_language_server")`. Will send PR soon with your proposal.

Definitely an option. Ideally, the language server could auto-install itself if it is not present.

_Originally posted by @yyoncho in https://github.com/emacs-lsp/lsp-mode/issues/539#issuecomment-447107262_

**Comment by bradyt** [comment]
@yyoncho I am not sure if I interpreted your meaning, but if I did, the following may be a first draft.

It first determines what the path to `dart_language_server` should look like, then if it's not there, it tries to find `pub` and install `dart_language_server`.

I know that some users are installing flutter and not the dart-sdk itself, will consider adding logic for that case as well. (We may need a similar case treatment in dart-mode... this lead to several issues.)

``` diff
@@ -291,9 +291,21 @@ finding the executable with `exec-path'."
 
 ;; Dart
 (defcustom lsp-clients-dart-server-command
-  (expand-file-name (if (equal system-type 'windows-nt)
-                        "~/Pub/Cache/bin/dart_language_server"
-                      "~/.pub-cache/bin/dart_language_server"))
+  (let ((dls (if (equal system-type 'windows-nt)
+                 (expand-file-name "~/Pub/Cache/bin/dart_language_server.bat")
+               (expand-file-name "~/.pub-cache/bin/dart_language_server"))))
+    (if (file-executable-p dls)
+        dls
+      (let ((pub (or (executable-find "pub")
+                     (when (file-executable-p "/usr/lib/dart/bin/pub")
+                       "/usr/lib/dart/bin/pub"))))
+        (if pub
+            (progn
+              (message "lsp-mode: Installing dart_language_server...")
+              (shell-command (concat pub " global activate dart_language_server"))
+              (message "lsp-mode: Installed dart_language_server...")
+              dls)
+          (error "lsp-mode: No pub executable found. Please install dart-sdk.")))))
   "The dart_language_server executable to use."
   :group 'lsp-dart
   :type 'file)
```

**Comment by yyoncho** [comment]
Looks good to me. Can you just move the installation to happen not in the defcustom but in a command before starting the server.  `lsp-stdio-connection` accepts also a function, e. g. `(lsp-stdio-connection 'lsp-groovy--lsp-command)` which will be invoked before starting the server.

> (We may need a similar case treatment in dart-mode... this lead to several issues.)

I don't have the story for the case when you do not have the major mode installed but lsp.el should handle it somehow. What kind of issues do you refer to, the code itself looks fine? If pub is not found, you could ask user to select the path to pub.

As a side note, there is also possibility to add a debugger support, check https://github.com/emacs-lsp/lsp-mode/pull/536#issuecomment-446899502 .

**Comment by yyoncho** [comment]
Fixed with #543
